client/my-sites/site-settings/theme-setup/index.jsx

```diff
@@ -14,41 +14,71 @@ import QueryActiveTheme from 'components/data/query-active-theme';
 import QueryTheme from 'components/data/query-theme';
 import ThemeSetupCard from './theme-setup-card';
 import ThemeSetupPlaceholder from './theme-setup-placeholder';
-import { getSelectedSite } from 'state/ui/selectors';
+import { getSelectedSiteId, getSelectedSiteSlug } from 'state/ui/selectors';
 import { getActiveTheme, getTheme } from 'state/themes/selectors';
+import { isJetpackSite } from 'state/sites/selectors';
 import { toggleDialog } from 'state/ui/theme-setup/actions';
 
-let ThemeSetup = ( { site, themeId, theme, translate, activeSiteDomain, toggleDialog } ) => {
-	const onBack = () => {
-		page( '/settings/general/' + activeSiteDomain );
+class ThemeSetup extends Component {
+	componentDidUpdate( prevProps ) {
+		const {
+			siteIsJetpack,
+			siteId,
+		} = this.props;
+
+		if ( siteId !== prevProps.siteId && siteIsJetpack ) {
+			this.redirectToGeneral();
+		}
+	}
+
+	redirectToGeneral = () => {
+		const { siteSlug } = this.props;
+		page.redirect( '/settings/general/' + siteSlug );
```

Comment:
Good catch Jon. 

Your suggestion wouldn't work though, as you'd still get a `/settings/general/null` if you click the link before the sites are loaded.

So I've done it in the `mapStateToProps` directly, see https://github.com/Automattic/wp-calypso/pull/15922/commits/076ebc090e27733e784d363fe01db3f2a6fc4976.